### PR TITLE
Issue19 distinct values

### PR DESCRIPTION
**Pull request by fkaufman-asym**

### CODE REVIEW

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 1]
```diff
@@ -89,4 +94,25 @@ private String createNextUrl(String jobId, int offset, int limit) {
...
+    Map<String, String> tableParts = QueryTranslator.parseTableName(body);
+    // SELECT * FROM UNNEST (SELECT DISTINCT p.column FROM qualifiedTABLE as p);
+    String querySql =
+        "SELECT DISTINCT "
+            + tableParts.get("columnName")
+            + " FROM "
+            + bqTable
+            + "."
+            + tableParts.get("tableName");
+    logger.info("uniqueValues: " + querySql);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:111]: This looks like a `debug` log, not `info`.

File: src/main/java/bio/terra/cda/app/util/QueryTranslator.java [hunk 2]
```diff
@@ -19,6 +20,15 @@ public static String sql(String table, Query query) {
     return new SqlGenerator(table, query).generate();
   }
 
+  /** Split the TABLE.ColumnName into component parts */
+  public static Map<String, String> parseTableName(String qualifiedTable) {
+    int dotPos = qualifiedTable.lastIndexOf('.');
+    String table = qualifiedTable.substring(0, dotPos);
+    String column = qualifiedTable.substring(dotPos + 1);
+
+    return Map.of("columnName", qualifiedTable.substring(dotPos + 1), "tableName", table);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/util/QueryTranslator.java:29]: can use `column` here
```suggestion
    return Map.of("columnName", column, "tableName", table);
```

File: src/main/java/bio/terra/cda/app/util/QueryTranslator.java [hunk 3]
```diff
@@ -19,6 +20,15 @@ public static String sql(String table, Query query) {
     return new SqlGenerator(table, query).generate();
   }
 
+  /** Split the TABLE.ColumnName into component parts */
+  public static Map<String, String> parseTableName(String qualifiedTable) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/util/QueryTranslator.java:24]: Having a public function that returns `Map` is not great since a caller will need to know what names the map contains and what type each element is. Either return a class here or split it into two methods. Or inline the code directly in the caller, as is done elsewhere in this class.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 4]
```diff
@@ -89,4 +94,25 @@ private String createNextUrl(String jobId, int offset, int limit) {
 
     return sendQuery(querySql, dryRun);
   }
+
+  @Override
+  public ResponseEntity<QueryResponseData> uniqueValues(String version, @Valid String body) {
+
+    String bqTable = "gdc-bq-sample";
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:101]: Instead of hardcoding this, you should be able to get it from `applicationConfiguration.getBqTable()`. It looks like `version` is ignored, so you need something like what's done in `bulkData()`:
```
    String querySql = "SELECT * FROM " + applicationConfiguration.getBqTable() + "." + version;
```

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 5]
```diff
@@ -89,4 +94,25 @@ private String createNextUrl(String jobId, int offset, int limit) {
...
+    Map<String, String> tableParts = QueryTranslator.parseTableName(body);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:102]: I think the body should be just the column name, not the table as well, as users shouldn't need to provide the table name.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 6]
```diff
@@ -89,4 +94,25 @@ private String createNextUrl(String jobId, int offset, int limit) {
 
     return sendQuery(querySql, dryRun);
   }
+
+  @Override
+  public ResponseEntity<QueryResponseData> uniqueValues(String version, @Valid String body) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:99]: Is `@Valid` needed here?

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 7]
```diff
@@ -89,4 +94,25 @@ private String createNextUrl(String jobId, int offset, int limit) {
...
+            + " FROM "
+            + bqTable
+            + "."
+            + tableParts.get("tableName");
+    logger.info("uniqueValues: " + querySql);
+
+    ResponseEntity<QueryCreatedData> response = sendQuery(querySql, false);
+    var query_id = response.getBody().getQueryId();
+
+    return query(query_id, 0, 1000);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:116]: Like the other query APIs, this should return `sendQuery(querySql, false)` directly. The API user will need to poll `query` for the result.

File: src/main/java/bio/terra/cda/app/service/QueryService.java [hunk 8]
```diff
@@ -232,4 +241,24 @@ public String startQuery(String query) {
 
     return queryJob.getJobId().getJob();
   }
+
+  /**
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/service/QueryService.java:245]: In general I would avoid committing commented out code, for many reasons. You can keep the code locally until it's ready to be used.

File: src/main/java/bio/terra/cda/app/service/QueryService.java [hunk 9]
```diff
@@ -155,6 +155,11 @@ private QueryResult getJobResults(Job queryJob, int offset, int pageSize) {
     }
   }
 
+  public boolean jobStatus(String jobId) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/service/QueryService.java:158]: This appears to be unused.

File: src/main/java/bio/terra/cda/app/service/QueryService.java [hunk 10]
```diff
@@ -223,6 +228,10 @@ private void logQuery(Job queryJob, List<JsonNode> jsonData) {
     }
   }
 
+  /**
+   * @param query
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/service/QueryService.java:232]: It's good to add javadoc but an empty javadoc should be deleted as it doesn't add anything to the code.

File: src/main/resources/api/service_openapi.yaml [hunk 11]
```diff
@@ -155,6 +155,28 @@ paths:
         200:
           $ref: '#/components/responses/QueryResponse'
 
+  /api/v1/unique-values/{version}:
+    post:
+      summary: Returns all distinct values
```
Comment (pshapiro4broad) [inline src/main/resources/api/service_openapi.yaml:160]: May as well use the same name in both the endpoint and the documentation. Either `unique` or `distinct` seems fine to me, as long as it's consistent.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 7 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java:116]: Alternatively, you could make this return the list of names and not return any of the query data at all. You'd have to change this code to loop over `query()` until it's complete, and then extract out the response body so it just contains a list of strings. If you want to do that, it would be better to do it in `QueryService` as its own method.

File: src/main/resources/api/service_openapi.yaml [hunk 12]
```diff
@@ -234,6 +257,9 @@ components:
     QueryResponseData:
       type: object
       properties:
+        query_id:
```
Comment (pshapiro4broad) [inline src/main/resources/api/service_openapi.yaml:260]: Why is this needed in the response? The caller already has it, since they needed it to make the API that returned it.

File: src/main/resources/api/service_openapi.yaml [hunk 13]
```diff
@@ -179,6 +201,7 @@ components:
       name: limit
       schema:
         type: integer
+        default: 1000
```
Comment (pshapiro4broad) [inline src/main/resources/api/service_openapi.yaml:204]: That's weird, I thought I had added that before. Thanks for fixing this.

File: src/main/resources/api/service_openapi.yaml [hunk 14]
```diff
@@ -155,6 +155,28 @@ paths:
         200:
           $ref: '#/components/responses/QueryResponse'
 
+  /api/v1/unique-values/{version}:
```
Comment (pshapiro4broad) [inline src/main/resources/api/service_openapi.yaml:158]: You need to update the version in `info` at the top when a new API is added. We haven't been good about this in the past, but we should be better about it in the future.

**Review by pshapiro4broad — CHANGES_REQUESTED** [review]

File: src/main/resources/api/service_openapi.yaml (hunk 11 above)
Comment (fkaufman-asym) [reply src/main/resources/api/service_openapi.yaml:160]: The original ticket asked for unique value,  distinct is probably more intuitive because that is the SQL call.

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 9 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:158]: Removed

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 6 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java]: Removed

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 5 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java]: That is what I'm leaving it as now.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 7 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java]: Now matches the other api calls.

File: src/main/resources/api/service_openapi.yaml (hunk 14 above)
Comment (fkaufman-asym) [reply src/main/resources/api/service_openapi.yaml:158]: incremented to 1.1.1

File: src/main/resources/api/service_openapi.yaml (hunk 12 above)
Comment (fkaufman-asym) [reply src/main/resources/api/service_openapi.yaml:260]: I took this back out for now.   My original thought was including as an anchor for maybe pasting between a programmatic output and back into the swagger.

File: src/main/java/bio/terra/cda/app/util/QueryTranslator.java (hunk 3 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/QueryTranslator.java]: I created a new model NestedColumn that will return an object.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 15]
```diff
@@ -95,24 +94,21 @@ private String createNextUrl(String jobId, int offset, int limit) {
     return sendQuery(querySql, dryRun);
   }
 
+  /**
+   * Table cells that represent as nested structures must be unnested from it decendant heiarchy.
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:98]: This duplicates the comment in `NestedColumn` so there's no need to put it here.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java [hunk 16]
```diff
@@ -95,24 +94,21 @@ private String createNextUrl(String jobId, int offset, int limit) {
...
-            + "."
-            + tableParts.get("tableName");
-    logger.info("uniqueValues: " + querySql);
-
-    ResponseEntity<QueryCreatedData> response = sendQuery(querySql, false);
-    var query_id = response.getBody().getQueryId();
-
-    return query(query_id, 0, 1000);
+    String table = applicationConfiguration.getBqTable() + "." + version;
+    NestedColumn nt = new bio.terra.cda.app.model.NestedColumn(body);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/QueryApiController.java:108]: Since this class is imported you don't need to use a fully qualified name.
```suggestion
    NestedColumn nt = new NestedColumn(body);
```

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Table cells that represent as nested structures must be unnested from it decendant heiarchy. the
+ * UNNEST clause is generated by unwinding the "dot" hierarchy such that:
+ * gdc-bg-sample.cda_mvp.v3.ResearchSubject.identifier.value would unwind to the following. "SELECT
+ * DISTINCT (_D.value) FROM gdc-bq-sample.cda_mvp.v3, UNNEST(ResearchSubject) as _C,
+ * UNNEST(_C.identifier) AS _D";
+ */
+public class NestedColumn {
+  private String column;
+  private String unnestClause;
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/model/NestedColumn.java:12]: Since these fields are constant after the constructor completes, it would be better to declare them as `final` to make this clear. To do this, you could move the code in `generateUnnestClause()` to the constructor. Another approach is to change this constructor to accept `column` and `unnestClause` and replace `generateUnnestClause()` with a static factory method that accepts `columnName` and returns a new `NestedColumn`.

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+  // column ->
+  // SELECT DISTINCT(column)
+
+  // D.column ->
+  // SELECT DISTINCT(_D.column) FROM TABLE, UNNEST(D) AS _D
+
+  // A.B.C.D.column ->
+  // # SELECT DISTINCT(_D.column) FROM TABLE, UNNEST(A) AS _A, UNNEST(_A.B) AS _B, UNNEST(_B.C) AS
+  // _C, UNNEST(_C.D) AS _D
+  private void generateUnnestClause(String columnName) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/model/NestedColumn.java:44]: This method is complex enough that it needs a unit test. You can unit test at the `NestedColumn()` level, e.g. create a `NestedColumn()` and assert that the `column` and `unnestClause` fields have the expected values.

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    String unnestClause = null;
+    String newColumn = null;
+    String[] c = columnName.split("\\.");
+    if (c.length > 1) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/model/NestedColumn.java:48]: Note that if `c.length` is less than one, the fields in this class will be `null`. Maybe this should throw an exception in that case?

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+      newColumn = "_" + c[c.length - 2] + "." + c[c.length - 1];
+      unnestClause = ", UNNEST(" + c[0] + ") AS _" + c[0];
+      for (int n = 1; n < c.length - 1; n++) {
+        unnestClause += ", UNNEST(_" + c[n - 1] + "." + c[n] + ") AS _" + c[n];
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/model/NestedColumn.java:52]: In general it's bad form to use `+=` on a `String` since that will reallocate the string every time it's used. If you have a string that you need to append to, it's usually best to use `StringBuilder`. https://docs.oracle.com/javase/tutorial/java/data/buffers.html

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 15 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java:98]: Removed duplicate comments

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 16 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java:108]: IDE did that...fixed.

File: src/main/java/bio/terra/cda/app/controller/QueryApiController.java (hunk 1 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/controller/QueryApiController.java]: fixed

File: src/main/java/bio/terra/cda/app/util/QueryTranslator.java (hunk 2 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/QueryTranslator.java]: refactored

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 8 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:240]: fixed

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 10 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:227]: fixed

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java (hunk 19 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/model/NestedColumn.java]: Handled this case by throwing an IllegalArgumentException

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java (hunk 18 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/model/NestedColumn.java]: Wrote a static Builder class that returns a NestedColumn object
That class has Unit testing implemented

File: src/main/java/bio/terra/cda/app/model/NestedColumn.java (hunk 20 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/model/NestedColumn.java]: Java-cized this bit of code.   Referenced the python version to get this working originally.

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        if (c.length > 1) {
+          newColumn = "_" + c[c.length - 2] + "." + c[c.length - 1];
+          unnestClause.append(", UNNEST(" + c[0] + ") AS _" + c[0]);
+          for (int n = 1; n < c.length - 1; n++) {
+            unnestClause.append(", UNNEST(_" + c[n - 1] + "." + c[n] + ") AS _" + c[n]);
+          }
+          return new NestedColumn(newColumn, unnestClause.toString());
+        } else if (c.length > 0) {
+          return new NestedColumn(qualifiedColumnName, "");
+        }
```
[the remark this answers is not in the document]
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/NestedColumn.java]: fixed

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+      }
+      // Case where a null or empty value is passed.
+      throw new IllegalArgumentException("");
```
[the remark this answers is not in the document]
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/NestedColumn.java]: You may be correct. I am not completely familiar with the OpenAPI validation yet.  I was originally aiming to protect against null pointer exceptions on the split if I did get passed a null.

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 10 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:227]: Marked as fixed but I don't see any changes in the PR diff

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 8 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:240]: I don't see any changes for this

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    this.column = column;
+    this.unnestClause = unnestClause;
+  }
+
+  public static NestedColumn generate(String qualifiedColumnName) {
+    return Builder.generate(qualifiedColumnName);
+  }
+
+  public static class Builder {
+    /**
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/util/NestedColumn.java:19]: This isn't a javadoc comment, so it shouldn't start with `/**`. You can use `//` or `/* ... */` forms for this.

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+        if (c.length > 1) {
+          newColumn = "_" + c[c.length - 2] + "." + c[c.length - 1];
+          unnestClause.append(", UNNEST(" + c[0] + ") AS _" + c[0]);
+          for (int n = 1; n < c.length - 1; n++) {
+            unnestClause.append(", UNNEST(_" + c[n - 1] + "." + c[n] + ") AS _" + c[n]);
+          }
+          return new NestedColumn(newColumn, unnestClause.toString());
+        } else if (c.length == 1) {
+          return new NestedColumn(qualifiedColumnName, "");
+        }
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/util/NestedColumn.java:42]: unnecessary `else` after `return`
```suggestion
        }
        if (c.length == 1) {
          return new NestedColumn(qualifiedColumnName, "");
        }
```

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+      // Case where a null or empty value is passed.
+      throw new IllegalArgumentException("Empty column name");
+    }
+  }
+
+  public String getColumn() {
+    return column;
+  }
+
+  public void setColumn(String column) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/util/NestedColumn.java:53]: This method and `setUnnestClause` don't appear to be used. If not, it would be better to remove them and change the fields `column` and `unnestClause` to be `final`.

File: src/main/resources/api/service_openapi.yaml [hunk 26]
```diff
@@ -155,6 +155,28 @@ paths:
...
+      description: Return unique values given a qualified columnName
+      operationId: uniqueValues
+      tags:
+        - query
+
+      parameters:
+        - $ref: '#/components/parameters/DatasetVersion'
+
+      requestBody:
+        description: TABLE.column_name of table value being requested
```
Comment (pshapiro4broad) [inline src/main/resources/api/service_openapi.yaml:170]: The input doesn't contain `TABLE`, it only contains column names, e.g., `column.column.column`.

File: src/test/java/bio/terra/cda/app/util/NestedColumnTest.java [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+    assertEquals(clause, result1.getUnnestClause());
+  }
+
+  @Test
+  public void testIllegalArgCondition() throws Exception {
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> {
+          NestedColumn.generate(null);
+        });
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/cda/app/util/NestedColumnTest.java:37]: Since the lambda body is a single line, you don't need `{` `}`. The IDE can fix this for you.
```suggestion
        () -> NestedColumn.generate(null));
```

File: src/test/java/bio/terra/cda/app/util/NestedColumnTest.java [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+    return Stream.of(
+        Arguments.of("A.B.C", "_B.C", ", UNNEST(A) AS _A, UNNEST(_A.B) AS _B"),
+        Arguments.of("A.B", "_A.B", ", UNNEST(A) AS _A"),
+        Arguments.of("A", "A", ""),
+        Arguments.of("", "", ""));
+  }
+
+  @ParameterizedTest
+  @MethodSource("unnestData")
+  public void testGeneratedUnnestClause2(String qualifiedName, String column, String clause)
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/cda/app/util/NestedColumnTest.java:24]: Not sure why this ends in `2`?
```suggestion
  public void testGeneratedUnnestClause(String qualifiedName, String column, String clause)
```

**Review by pshapiro4broad — CHANGES_REQUESTED** [review]
A few more minor changes, and some (apparently) unresolved comments from before.

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java (hunk 23 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/NestedColumn.java:19]: Changed to multiline comment

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java (hunk 24 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/NestedColumn.java:42]: Fixed

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java (hunk 25 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/NestedColumn.java:53]: Removed these unused setters.  Also removed the public constructor since there is only the static generate method.

File: src/main/resources/api/service_openapi.yaml (hunk 26 above)
Comment (fkaufman-asym) [reply src/main/resources/api/service_openapi.yaml:170]: FIxed this.   
Originally thought we would need inspection of tables beyond the cda_mvp

File: src/test/java/bio/terra/cda/app/util/NestedColumnTest.java (hunk 28 above)
Comment (fkaufman-asym) [reply src/test/java/bio/terra/cda/app/util/NestedColumnTest.java:24]: cleaned this code up.

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 10 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:227]: There was some commented out code that somehow got back in.   I removed it again.

**Comment by fkaufman-asym** [comment]
I think I have everything Phil commented on.   This will close out the last of the tickets that will be included in the next tagged release.

**Comment by pshapiro4broad** [comment]
> I think I have everything Phil commented on. This will close out the last of the tickets that will be included in the next tagged release.

There are still a number of comments that you've replied Fixed to but no changes have been pushed. Did you forget to push your changes from your local branch?

**Comment by fkaufman-asym** [comment]
Sorry,  Just pushed.

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 10 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:227]: This is still here, but empty. Please either delete or add javadoc to the comment.

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+    this.unnestClause = unnestClause;
+  }
+
+  public static NestedColumn generate(String qualifiedColumnName) {
+    return Builder.generate(qualifiedColumnName);
+  }
+
+  public static class Builder {
+    /*
+     column -> SELECT DISTINCT(column)
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/util/NestedColumn.java:18]: This is better but still not correctly formatted. In google code style, block comments start with `/*` and end with `*/` and have `* ` on each line.  https://google.github.io/styleguide/javaguide.html#s4.8.6-comments

File: src/main/resources/api/service_openapi.yaml (hunk 12 above)
Comment (pshapiro4broad) [reply src/main/resources/api/service_openapi.yaml:260]: Is this taken out or not? It looks like it's still here in the in the yaml.

**Review by pshapiro4broad — CHANGES_REQUESTED** [review]
Almost done, just a one or two things left to fix.

File: src/main/java/bio/terra/cda/app/util/NestedColumn.java (hunk 29 above)
Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/util/NestedColumn.java]: formatted according to style guide.

**Review by pshapiro4broad — APPROVED** [review]
Looks good!

**Comment by fkaufman-asym** [comment]
Fixes #19

**Comment by fkaufman-asym** [comment]
Fixes #19
